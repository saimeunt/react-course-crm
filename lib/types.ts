import { z } from "zod";

const customerSchema = z.object({
  id: z.number().min(1),
  firstName: z.string().min(1),
  lastName: z.string().min(1),
  email: z.string().email(),
  profileUrl: z.string().url(),
  createdAt: z.date(),
  updatedAt: z.date(),
  deletedAt: z.date().nullable(),
  fullName: z.string(),
  totalSpent: z.number().min(0),
});

export type Customer = z.infer<typeof customerSchema>;

export const customerSubSchema = customerSchema.pick({
  id: true,
  firstName: true,
  lastName: true,
  fullName: true,
  profileUrl: true,
  deletedAt: true,
});

export const upsertCustomerFormSchema = customerSchema.pick({
  firstName: true,
  lastName: true,
  email: true,
});

export const productSchema = z.object({
  id: z.number().min(1),
  name: z.string().min(1),
  description: z.string(),
  price: z.number().min(1).max(1000).multipleOf(0.01),
  imageUrl: z.string().url(),
});

export type Product = z.infer<typeof productSchema>;

export const productSubSchema = productSchema.pick({
  id: true,
  name: true,
  price: true,
  imageUrl: true,
});

export const upsertProductFormSchema = productSchema.pick({
  name: true,
  description: true,
  price: true,
  imageUrl: true,
});

export const orderProductSchema = z.object({
  productId: z.number().min(1),
  product: productSubSchema,
  quantity: z.number().min(1).max(10),
});

export const orderStatusKeys = [
  "IN_PROGRESS",
  "COMPLETED",
  "CANCELLED",
] as const;

export type OrderStatus = (typeof orderStatusKeys)[number];

export const orderStatuses: Record<OrderStatus, string> = {
  IN_PROGRESS: "In progress",
  COMPLETED: "Completed",
  CANCELLED: "Cancelled",
} as const;

type BadgeVariant =
  | "default"
  | "secondary"
  | "destructive"
  | "outline"
  | null
  | undefined;

export const orderStatusesBadgeVariants: Record<OrderStatus, BadgeVariant> = {
  IN_PROGRESS: "outline",
  COMPLETED: "default",
  CANCELLED: "destructive",
} as const;

export const ORDER_MIN_PRODUCTS = 1;
export const ORDER_MAX_PRODUCTS = 5;

export const orderSchema = z.object({
  id: z.number().min(1),
  customerId: z.number().min(1),
  customer: customerSubSchema,
  status: z.enum(orderStatusKeys),
  createdAt: z.date(),
  updatedAt: z.date(),
  deletedAt: z.date().nullable(),
  products: z
    .array(orderProductSchema)
    .min(ORDER_MIN_PRODUCTS)
    .max(ORDER_MAX_PRODUCTS),
  totalPrice: z.number().min(0),
});

export type Order = z.infer<typeof orderSchema>;

export const upsertOrderFormSchema = orderSchema
  .pick({
    customerId: true,
    status: true,
  })
  .extend({
    products: z
      .array(
        orderProductSchema.pick({
          productId: true,
          quantity: true,
        }),
      )
      .min(1)
      .max(5),
  });
