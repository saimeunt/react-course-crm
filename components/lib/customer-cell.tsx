import { z } from "zod";
import { Avatar, AvatarFallback, AvatarImage } from "@/components/ui/avatar";
import { customerSubSchema } from "@/lib/types";
import { CircleUser } from "lucide-react";

const CustomerCell = ({
  customer,
}: {
  customer: z.infer<typeof customerSubSchema>;
}) => (
  <div className="flex items-center gap-2">
    {customer.deletedAt ? (
      <>
        <CircleUser className="size-8" stroke="var(--color-destructive)" />
        <span className="text-destructive">Deleted customer</span>
      </>
    ) : (
      <>
        <Avatar>
          <AvatarImage
            src={customer.profileUrl}
            alt={`${customer.fullName} profile`}
          />
          <AvatarFallback>
            {customer.firstName.at(0)}
            {customer.lastName.at(0)}
          </AvatarFallback>
        </Avatar>
        <span>{customer.fullName}</span>
      </>
    )}
  </div>
);

export default CustomerCell;
