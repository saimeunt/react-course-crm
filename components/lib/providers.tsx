"use client";
import { type ReactNode } from "react";
import ThemeProvider from "@/components/lib/theme-provider";
import GlobalContextProvider from "@/components/global-context/provider";

const Providers = ({ children }: { children: ReactNode }) => (
  <ThemeProvider
    attribute="class"
    defaultTheme="system"
    enableSystem
    disableTransitionOnChange
    enableColorScheme
  >
    <GlobalContextProvider>{children}</GlobalContextProvider>
  </ThemeProvider>
);

export default Providers;
