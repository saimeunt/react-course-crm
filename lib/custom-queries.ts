import { queryHelpers, buildQueries } from "@testing-library/react";

// The queryAllByAttribute is a shortcut for attribute-based matchers
// You can also use document.querySelector or a combination of existing
// testing library utilities to find matching nodes for your query
const queryAllByDataSlot = (container: HTMLElement, slot: string) =>
  queryHelpers.queryAllByAttribute("data-slot", container, slot);

const getMultipleError = (container: Element | null, dataSlotValue: string) =>
  `Found multiple elements with the data-cy attribute of: ${dataSlotValue}`;
const getMissingError = (container: Element | null, dataSlotValue: string) =>
  `Unable to find an element with the data-cy attribute of: ${dataSlotValue}`;

const [
  queryByDataSlot,
  getAllByDataSlot,
  getByDataSlot,
  findAllByDataSlot,
  findByDataSlot,
] = buildQueries(queryAllByDataSlot, getMultipleError, getMissingError);

export {
  queryByDataSlot,
  getAllByDataSlot,
  getByDataSlot,
  findAllByDataSlot,
  findByDataSlot,
};
