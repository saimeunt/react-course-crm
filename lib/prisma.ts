//import { PrismaBetterSQLite3 } from "@prisma/adapter-better-sqlite3";
// import { PrismaPg } from "@prisma/adapter-pg";
import { withAccelerate } from "@prisma/extension-accelerate";
import { PrismaClient } from "@/prisma/generated/client";

const globalForPrisma = globalThis as unknown as { prisma: PrismaClient };

export const prisma =
  globalForPrisma.prisma || new PrismaClient().$extends(withAccelerate());
