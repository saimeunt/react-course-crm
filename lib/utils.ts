import { clsx, type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export const cn = (...inputs: ClassValue[]) => twMerge(clsx(inputs));

export const formatPrice = (price: number) =>
  new Intl.NumberFormat("en-US", { style: "currency", currency: "USD" }).format(
    price,
  );

export const formatPercent = (
  percent: number,
  signDisplay: "auto" | "never" | "always" | "exceptZero" = "auto",
) =>
  new Intl.NumberFormat("en-US", {
    style: "percent",
    maximumFractionDigits: 2,
    signDisplay,
  }).format(percent);

export const formatNumber = (
  value: number | bigint,
  options?: Intl.NumberFormatOptions,
) => new Intl.NumberFormat("en-US", options).format(value);

export const fileToBase64 = (file: File) =>
  new Promise<string>((resolve) => {
    const fileReader = new FileReader();
    fileReader.addEventListener(
      "load",
      () => resolve(fileReader.result as string),
      false,
    );
    fileReader.readAsDataURL(file);
  });
