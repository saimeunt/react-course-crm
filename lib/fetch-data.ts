import { cache } from "react";
import { prisma } from "@/lib/prisma";

export const fetchCustomers = async () => {
  const customers = await prisma.customer.findMany({
    where: { deletedAt: null },
    orderBy: { createdAt: "desc" },
  });
  return customers.map((customer) => ({
    ...customer,
    fullName: `${customer.firstName} ${customer.lastName}`,
    totalSpent: 0,
  }));
};

export const fetchCustomersWithTotalSpent = async () => {
  const customers = await prisma.customer.findMany({
    where: { deletedAt: null },
    orderBy: { createdAt: "desc" },
    include: {
      orders: {
        include: {
          products: {
            select: {
              product: {
                select: { price: true },
              },
              quantity: true,
            },
          },
        },
      },
    },
  });
  return customers.map((customer) => ({
    ...customer,
    fullName: `${customer.firstName} ${customer.lastName}`,
    totalSpent: customer.orders.reduce(
      (previousValue, order) =>
        previousValue +
        order.products.reduce(
          (previousValue, { product, quantity }) =>
            previousValue + product.price * quantity,
          0,
        ),
      0,
    ),
  }));
};

export const fetchCustomer = cache(async (id: number) => {
  const customer = await prisma.customer.findUnique({
    where: { id, deletedAt: null },
  });
  if (!customer) {
    return null;
  }
  return {
    ...customer,
    fullName: `${customer.firstName} ${customer.lastName}`,
    totalSpent: 0,
  };
});

export const fetchOrders = async () => {
  const orders = await prisma.order.findMany({
    where: { deletedAt: null },
    orderBy: { createdAt: "desc" },
    include: {
      customer: {
        select: {
          id: true,
          firstName: true,
          lastName: true,
          profileUrl: true,
          deletedAt: true,
        },
      },
      products: {
        select: {
          product: {
            select: { id: true, name: true, price: true, imageUrl: true },
          },
          productId: true,
          quantity: true,
        },
      },
    },
  });
  return orders.map((order) => ({
    ...order,
    customer: {
      ...order.customer,
      fullName: `${order.customer.firstName} ${order.customer.lastName}`,
    },
    totalPrice: order.products.reduce(
      (previousValue, { product, quantity }) =>
        previousValue + product.price * quantity,
      0,
    ),
  }));
};

export const fetchOrder = async (id: number) => {
  const order = await prisma.order.findUnique({
    where: { id },
    include: {
      customer: {
        select: {
          id: true,
          firstName: true,
          lastName: true,
          profileUrl: true,
          deletedAt: true,
        },
      },
      products: {
        select: {
          product: {
            select: { id: true, name: true, price: true, imageUrl: true },
          },
          productId: true,
          quantity: true,
        },
      },
    },
  });
  if (!order) {
    return null;
  }
  return {
    ...order,
    customer: {
      ...order.customer,
      fullName: `${order.customer.firstName} ${order.customer.lastName}`,
    },
    totalPrice: order.products.reduce(
      (previousValue, { product, quantity }) =>
        previousValue + product.price * quantity,
      0,
    ),
  };
};

export const fetchProducts = async () =>
  prisma.product.findMany({
    where: { deletedAt: null },
    orderBy: { createdAt: "desc" },
    select: {
      id: true,
      name: true,
      description: true,
      price: true,
      imageUrl: true,
    },
  });
