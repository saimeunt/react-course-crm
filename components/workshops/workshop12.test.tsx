/**
 * @jest-environment jsdom
 */
import { render, queries, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import * as customQueries from "@/lib/custom-queries";
import Workshop12, {
  TogglableCounterControlled,
} from "@/components/workshops/workshop12";

const allQueries = {
  ...queries,
  ...customQueries,
};

const screen = within(document.body, allQueries);

describe("Workshop12", () => {
  it("renders Workshop12 unchanged", () => {
    const { container } = render(<Workshop12 />);
    expect(container).toMatchSnapshot();
  });
});

describe("TogglableCounterControlled", () => {
  beforeEach(async () => {
    // ARRANGE
    render(<TogglableCounterControlled />);
    // ACT
    await userEvent.click(screen.getByRole("switch"));
    // ASSERT
  });
  it("has an initial value of 0", () => {
    // ARRANGE
    // ACT
    // ASSERT
    expect(screen.getByDataSlot("card-title")).toHaveTextContent("0");
  });
  it("decreases the counter when clicking the Decrease button", async () => {
    // ARRANGE
    // ACT
    await userEvent.click(screen.getByText("Decrease"));
    // ASSERT
    expect(screen.getByDataSlot("card-title")).toHaveTextContent("-1");
  });
  it("increases the counter when clicking the Increase button", async () => {
    // ARRANGE
    // ACT
    await userEvent.click(screen.getByText("Increase"));
    // ASSERT
    expect(screen.getByDataSlot("card-title")).toHaveTextContent("1");
  });
});
