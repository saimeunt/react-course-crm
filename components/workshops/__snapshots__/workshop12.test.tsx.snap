// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workshop12 renders Workshop12 unchanged 1`] = `
<div>
  <div
    class="flex flex-col px-4 lg:px-6 gap-6"
  >
    <div
      class="flex flex-col items-center gap-4"
    >
      <div
        class="flex flex-col items-center gap-4"
      >
        <div
          class="flex items-center space-x-2"
        >
          <button
            aria-checked="false"
            class="peer data-[state=checked]:bg-primary data-[state=unchecked]:bg-input focus-visible:border-ring focus-visible:ring-ring/50 dark:data-[state=unchecked]:bg-input/80 inline-flex h-[1.15rem] w-8 shrink-0 items-center rounded-full border border-transparent shadow-xs transition-all outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50"
            data-slot="switch"
            data-state="unchecked"
            id="toggle-counter-controlled"
            role="switch"
            type="button"
            value="on"
          >
            <span
              class="bg-background dark:data-[state=unchecked]:bg-foreground dark:data-[state=checked]:bg-primary-foreground pointer-events-none block size-4 rounded-full ring-0 transition-transform data-[state=checked]:translate-x-[calc(100%-2px)] data-[state=unchecked]:translate-x-0"
              data-slot="switch-thumb"
              data-state="unchecked"
            />
          </button>
          <label
            class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50"
            data-slot="label"
            for="toggle-counter-controlled"
          >
            Show
             Counter
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
